### PR TITLE
wth - Refactor add_service method in

### PR DESCRIPTION
**Pull request by wtholt**
app/controllers/service_requests_controller.rb

Refactored add_service method, extracted logic into ruby class.

[#121053875]

Story - https://www.pivotaltracker.com/story/show/121053875

### CODE REVIEW

File: app/controllers/service_requests_controller.rb [hunk 1]
```diff
@@ -188,29 +188,15 @@ def save_and_exit
   end
 
   def add_service
-    existing_service_ids = @service_request.line_items.reject{ |line_item| FINISHED_STATUSES.include?(line_item.status) }.map(&:service_id)
-
-    if existing_service_ids.include?( params[:service_id].to_i )
+    add_service = AddService.new(@service_request,
```
Comment (kyle-glick) [inline app/controllers/service_requests_controller.rb]: While I think it's good to keep the controller method a bit shorter by extracting to a lib file, I think this needs more work to actually improve upon the existing code. This feels like a band-aid more than a fix.

The biggest problem I've had with this controller action is that it's _really slow_ and there's a lot of extra work being done and poor optimization that needs to be improved.

**Review by SamWord — CHANGES_REQUESTED** [review]
Why move those private methods to a new lib file? It makes the file look shorter but it might make `add_service` harder to follow if a bug were to arise.  Generally I would make them private methods in the controller with descriptive names.

**Comment by jleonardw9** [comment]
@SamWord It's pretty standard practice to refactor fat controllers in Rails using service objects. This is done using a PORO, or Plain Old Ruby Object (class), like Will has done here. Check out this article for more info: http://www.thegreatcodeadventure.com/service-objects-in-rails/

**Review by jleonardw9 — APPROVED** [review]

**Comment by wtholt** [comment]
@SamWord This is a standard way of refactoring in OOP. 

As this helpful guide states - 

- This refactoring method will help maintain adherence to the Single Responsibility Principle. The code of your classes will be more obvious and understandable.

- Single-responsibility classes are more reliable and tolerant of changes. For example, say that you have a class responsible for ten different things. When you change this class to make it better for one thing, you risk breaking it for the nine others.

Reference - https://sourcemaking.com/refactoring/extract-class

**Comment by amcates** [comment]
Looks great to me and makes perfect sense to move the logic out to a base class.   Rejecting a pull request should be reserved for when the code doesn't function as it should or would create a bug.  Otherwise, offer up an alternate solution offline.

**Review by SamWord — APPROVED** [review]

**Pull request closed without merging**